integratedauth/krb5/krb5.go

```diff
@@ -0,0 +1,268 @@
+		params3 = strings.Split(params3[1], ".")
+		realm = params3[1] + "." + params3[2]
+	case 2:
+		realm = params3[1]
+		serviceStr = params3[0]
+	default:
+		return nil, errors.New("invalid ServerSPN")
+	}
+
+	return &krb5Auth{
+		username:   config.User,
+		password:   config.Password,
+		serverSPN:  serviceStr,
+		port:       port,
+		realm:      realm,
+		krb5Config: krb.Config,
+		krbKeytab:  krb.Keytab,
+		krbCache:   krb.Cache,
+	}, nil
+}
+
+func (auth *krb5Auth) InitialBytes() ([]byte, error) {
+	var cl *client.Client
+	var err error
+	// Init keytab from conf
+	if auth.username != "" && auth.password != "" {
+		cl = client.NewWithPassword(auth.username, auth.realm, auth.password, auth.krb5Config)
+	} else if auth.krbKeytab != nil {
+		// Init krb5 client and login
+		cl = client.NewWithKeytab(auth.username, auth.realm, auth.krbKeytab, auth.krb5Config, client.DisablePAFXFAST(true))
+	} else {
+		cl, err = client.NewFromCCache(auth.krbCache, auth.krb5Config)
+		if err != nil {
+			return []byte{}, err
+		}
+	}
+	auth.krb5Client = cl
+	auth.state = initiatorStart
+	tkt, sessionKey, err := cl.GetServiceTicket(auth.serverSPN)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	negTok, err := spnego.NewNegTokenInitKRB5(auth.krb5Client, tkt, sessionKey)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	outToken, err := negTok.Marshal()
+	if err != nil {
+		return []byte{}, err
+	}
+	auth.state = initiatorWaitForMutal
+	return outToken, nil
+}
+
+func (auth *krb5Auth) Free() {
+	auth.krb5Client.Destroy()
+}
+
+func (auth *krb5Auth) NextBytes(token []byte) ([]byte, error) {
+	var spnegoToken spnego.SPNEGOToken
+	if err := spnegoToken.Unmarshal(token); err != nil {
+		err := fmt.Errorf("unmarshal APRep token failed: %w", err)
+		return []byte{}, err
+	}
+	auth.state = initiatorReady
+	return []byte{}, nil
+}
+
+func readKrb5Config(config msdsn.Config) (Kerberos, error) {
+	krb := Kerberos{}
+	var err error
+
+	krbConfig, ok := config.Parameters["krb5conffile"]
+	if !ok {
+		return krb, fmt.Errorf("krb5 config file is required")
+	}
+
+	krb.Config, err = SetKrbConfig(krbConfig)
+	if err != nil {
+		return krb, err
+	}
+
+	missingParam := validateKerbConfig(config.Parameters)
+	if missingParam != "" {
+		return krb, fmt.Errorf("missing parameter:%s", missingParam)
+	}
+
+	if realm, ok := config.Parameters["realm"]; ok {
+		krb.Realm = realm
+	}
+
+	if krbCache, ok := config.Parameters["krbcache"]; ok {
+		krb.Cache, err = SetKrbCache(krbCache)
+		if err != nil {
+			return krb, err
+		}
+	}
+
+	if keytabfile, ok := config.Parameters["keytabfile"]; ok {
+		krb.Keytab, err = SetKrbKeytab(keytabfile)
+		if err != nil {
+			return krb, err
+		}
+	}
+
+	return krb, nil
+}
+
+func validateKerbConfig(c map[string]string) (missingParam string) {
+	if c["keytabfile"] != "" {
+		if c["realm"] == "" {
+			missingParam = "realm"
+			return
+		}
+	}
+	if c["krbcache"] == "" && c["keytabfile"] == "" {
+		missingParam = "atleast krbcache or keytab is required"
```

Comment:
yes, there's a default location for keytab and krbcache at /etc/krb5/<file> and /tmp/krb5cc_%{uid} respectively. 

the reason we didn't look there in the first place was that the user will be aware of the exact keytab file being used just in case their system has multiple. 

do we want to check this location in case the user hasnt provided one ? can add this support.